packages/block-editor/src/components/inserter/block-patterns.js

```diff
@@ -73,29 +94,32 @@ function BlockPatterns( { patterns, onInsert, filterValue } ) {
 		);
 	}, [] );
 
-	return !! filteredPatterns.length ? (
+	if ( filterValue ) {
+		return !! filteredPatterns.length ? (
+			<InserterPanel title={ __( 'Search Results' ) }>
+				<BlockPatternList
+					patterns={ filteredPatterns }
+					onClickPattern={ onClickPattern }
+				/>
+			</InserterPanel>
+		) : (
+			<InserterNoResults />
+		);
+	}
+
+	return blockPatternCategories.map( ( patternCategory ) => (
 		<InserterPanel
-			title={
-				filterValue
-					? __( 'Search Results' )
-					: _x( 'All', 'patterns categories' )
-			}
+			key={ patternCategory.name }
+			title={ patternCategory.label }
 		>
-			{ filteredPatterns.map( ( pattern, index ) =>
-				currentShownPatterns[ index ] === pattern ? (
-					<BlockPattern
-						key={ pattern.name }
-						pattern={ pattern }
-						onClick={ onClickPattern }
-					/>
-				) : (
-					<BlockPatternPlaceholder key={ pattern.name } />
-				)
-			) }
+			<BlockPatternList
+				patterns={ patterns.filter( ( pattern ) =>
```

Comment:
This will show the category panel even if there are no categories in it. I'll fix that.